tools/tfs-fuse.c

```diff
@@ -0,0 +1,1013 @@
+    pthread_rwlock_unlock(&rwlock);
+    return rv;
+}
+
+/* The target provided to symlink is provided as-is, and not
+ * made relative to the mount point of the fs as paths in other
+ * callbacks are. This normally makes sense since the target could be
+ * anywhere, but not for tfs where a volume is normally used in a vm.
+ * The best implementation of this callback should involve sanitizing
+ * and transforming the target path into a relative path between files
+ * on the same volume or perhaps an absolute path from the root
+ * of the volume (not the host root). Not using the callback seems
+ * like the best choice to avoid mis-use or errors in symlinking. */
+#if 0
+static int tfs_symlink(const char *target, const char *path)
+{
+    tfs_fuse_debug("%s: path %s target %s\n", __func__, path, target);
+    filesystem fs = rootfs;
+    tuple parent;
+    pthread_rwlock_wrlock(&rwlock);
+    int rv = resolve_cstring(&fs, cwd, path, 0, &parent);
+    if ((rv != -ENOENT) || !parent) {
+        tfs_fuse_debug("symlink ret %d parent %p\n", ret, parent);
+        goto out;
+    }
+    if (filesystem_symlink(fs, parent, filename_from_path(path),
+        target)) {
+        rv = 0;
+        set_flush_timeout();
+    } else
+        rv = -ENOSPC;
+out:
+    pthread_rwlock_unlock(&rwlock);
+    return rv;
+}
+#endif
+
+static int statfs_internal(filesystem fs, tuple t, struct statvfs *buf)
+{
+    if (!buf) {
+        return -EFAULT;
+    }
+    runtime_memset((u8 *) buf, 0, sizeof(*buf));
+    if (fs) {
+        buf->f_bsize = fs_blocksize(fs);
+        buf->f_blocks = fs_totalblocks(fs);
+        buf->f_bfree = buf->f_bavail = fs_freeblocks(fs);
+    } else {
+        buf->f_bsize = PAGESIZE;
+    }
+    buf->f_frsize = buf->f_bsize;
+    u64 id = u64_from_pointer(t);
+    buf->f_fsid = id;
+    buf->f_namemax = NAME_MAX;
+    return 0;
+}
+
+static int tfs_statfs(const char *path, struct statvfs *buf)
+{
+    filesystem fs = rootfs;
+    tuple t;
+    pthread_rwlock_wrlock(&rwlock);
+    int rv = resolve_cstring(&fs, cwd, path, &t, 0);
+    if (rv)
+        goto out;
+    rv = statfs_internal(fs, t, buf);
+out:
+    pthread_rwlock_unlock(&rwlock);
+    return rv;
+}
+
+static int utime_internal(const char *filename, timestamp actime,
+        timestamp modtime)
+{
+    tuple t;
+    filesystem fs = rootfs;
+    int ret = resolve_cstring(&fs, cwd, filename, &t, 0);
+    if (ret) {
+        return ret;
+    }
+    filesystem_set_atime(fs, t, actime);
+    filesystem_set_mtime(fs, t, modtime);
+    return 0;
+}
+
+static inline timestamp time_from_timespec(const struct timespec *t)
+{
+    return seconds(t->tv_sec) + nanoseconds(t->tv_nsec);
+}
+
+static int tfs_utimens(const char *filename, const struct timespec tv[2])
+{
+    timestamp atime =
+        tv ? time_from_timespec(&tv[0]) : now(CLOCK_ID_REALTIME);
+    timestamp mtime =
+        tv ? time_from_timespec(&tv[1]) : now(CLOCK_ID_REALTIME);
+    return utime_internal(filename, atime, mtime);
```

Comment:
I think before calling utime_internal() it should call pthread_rwlock_wrlock(&rwlock)

Reply:
Yes, good catch. The modification of the metadata needs to be protected.